hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -0,0 +1,208 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.SqlCloseCodec;
+import com.hazelcast.client.impl.protocol.codec.SqlExecuteCodec;
+import com.hazelcast.client.impl.protocol.codec.SqlFetchCodec;
+import com.hazelcast.client.impl.spi.impl.ClientInvocation;
+import com.hazelcast.client.impl.spi.impl.ClientInvocationFuture;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.SqlException;
+import com.hazelcast.sql.SqlQuery;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.QueryUtils;
+
+import javax.annotation.Nonnull;
+import java.security.AccessControlException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Client-side implementation of SQL service.
+ */
+public class SqlClientService implements SqlService {
+
+    private final HazelcastClientInstanceImpl client;
+
+    public SqlClientService(HazelcastClientInstanceImpl client) {
+        this.client = client;
+    }
+
+    @Nonnull
+    @Override
+    public SqlResult query(@Nonnull SqlQuery query) {
+        Connection connection = client.getConnectionManager().getRandomConnection(true);
+
+        if (connection == null) {
+            throw rethrow(QueryException.error(
+                SqlErrorCode.CONNECTION_PROBLEM,
+                "Client must be connected to at least one data member to execute SQL queries"
+            ));
+        }
+
+        try {
+            List<Object> params = query.getParameters();
+
+            List<Data> params0 = new ArrayList<>(params.size());
+
+            for (Object param : params) {
+                params0.add(serializeParameter(param));
+            }
+
+            ClientMessage requestMessage = SqlExecuteCodec.encodeRequest(
+                query.getSql(),
+                params0,
+                query.getTimeoutMillis(),
+                query.getCursorBufferSize()
+            );
+
+            ClientMessage responseMessage = invoke(requestMessage, connection);
+
+            SqlExecuteCodec.ResponseParameters responseParameters = SqlExecuteCodec.decodeResponse(responseMessage);
+
+            SqlExecuteResponse response = new SqlExecuteResponse(
```

Comment:
There is no need for `SqlExecuteResponse` object anymore here.

Reply:
Fixed.